scipy/spatial/src/distance_metrics.h

```diff
@@ -0,0 +1,128 @@
+#pragma once
+
+#include <cmath>
+#include "views.h"
+
+struct MinkowskiDistance {
+    double p_;
+
+    template <typename T>
+    void operator()(StridedView2D<T> out, StridedView2D<const T> x, StridedView2D<const T> y) const {
+        T p = p_;
+        for (intptr_t i = 0; i < x.shape[0]; ++i) {
+            T p_dist = 0;
+            for (intptr_t j = 0; j < x.shape[1]; ++j) {
+                auto diff = std::abs(x(i, j) - y(i, j));
+                p_dist += std::pow(diff, p);
```

Comment:
Putting this general comment/question here.
Would it improve speed if we would just do the calculations for the lower (or upper) triangular matrix and then fill the remaining elements by symmetry?

Reply:
To produce a dense array it would involve a transposition of the triangular, which would incure strided memory access. With a modern CPU it is hard to say which will be better. It might also depend on the size of the array, as random access to L1, L2 and RAM has very different delays. The proof of the pudding is in the eating. 

It might be tempting to construct a class where inte indexing operations ```(i,j)``` and ```(j,i)``` references the same data. It just requires a conditional swap of ```i``` and ```j``` depending on which is larger. That would elide (or at least lazy evaluate) the transposition. However, it would not be particularly NumPy friendly. I wonder if it might be possible to produce the same effect in a strided NumPy array using some stride trick.

Reply:
As a rough way of estimating this, cityblock of 1000 elements in the benchmark takes 1.18 ms for `cdist` and 0.60 ms for `pdist`. Transpose-copying a 1000x1000 array also takes 0.51 ms in NumPy. So in this example, `pdist` + transpose is 1.11 ms and `cdist` is 1.18 ms.

The more time consuming the metric calculation; the more worth-while this optimization is likely to be. So it's something I'll bear in mind, but will leave for another PR.

Reply:
Thanks for looking into this 😄  Sure this is for another PR. I just did not want to open an issue for this comment.